File: src/providers/kimi/auth/device-id.ts
import { mkdir, readFile, writeFile, chmod } from "node:fs/promises";
import { dirname, join } from "node:path";
import { kimiDeviceIdFile, legacyConfigDir } from "../../../paths.ts";

function path(): string {
  return kimiDeviceIdFile();
}
function legacyPath(): string {
  return join(legacyConfigDir(), "kimi", "device_id");
}

export async function getDeviceId(): Promise<string> {
  for (const candidate of [path(), legacyPath()]) {
    try {
      const raw = await readFile(candidate, "utf8");
      const trimmed = raw.trim();
      if (trimmed) return trimmed;
    } catch (err: any) {
      if (err?.code !== "ENOENT") throw err;
    }
  }
  const id = crypto.randomUUID().replace(/-/g, "");
  const target = path();
  await mkdir(dirname(target), { recursive: true });
  await writeFile(target, id, { encoding: "utf8", mode: 0o600 });
  try {
    await chmod(target, 0o600);
  } catch {}
  return id;
}

File: src/providers/kimi/auth/device_id.rs
use std::fs;
#[cfg(unix)]
use std::path::PathBuf;

use crate::paths;

fn device_id_path() -> PathBuf {
    let deps = paths::DirResolverEnv::default();
    paths::kimi_device_id_file(&deps)
}

fn legacy_device_id_path() -> PathBuf {
    let home = std::env::var("HOME")
        .or_else(|_| std::env::var("USERPROFILE"))
        .unwrap_or_else(|_| "/".to_string());
    std::path::Path::new(&home)
        .join(".config")
        .join("claude-code-proxy")
        .join("kimi")
        .join("device_id")
}

pub fn get_device_id() -> Result<String, anyhow::Error> {
    for candidate in [device_id_path(), legacy_device_id_path()] {
        if let Ok(raw) = fs::read_to_string(&candidate) {
            let trimmed = raw.trim().to_string();
            if !trimmed.is_empty() {
                return Ok(trimmed);
            }
        }
    }

    let id = uuid::Uuid::new_v4().to_string().replace('-', "");
    let target = device_id_path();
    if let Some(dir) = target.parent() {
        fs::create_dir_all(dir)?;
        crate::paths::set_mode(dir, 0o700);
    }
    fs::write(&target, &id)?;
    crate::paths::set_mode(&target, 0o600);
    Ok(id)
}
